performance/multi-node/random_access.c

```diff
@@ -0,0 +1,1081 @@
+ * Likewise, it may be required to define a constraint as to the number of
+ * update messages that can be stored before processing to permit multi-level
+ * parallelism for those systems that support such a paradigm.  The limits on
+ * "look ahead" and "stored updates" are being implemented to assure that the
+ * benchmark meets the intent to profile memory architecture and not induce
+ * significant artificial data locality. For the purpose of measuring GUPS,
+ * we will stipulate that each thread is permitted to look ahead no more than
+ * 1024 random address stream samples with the same number of update messages
+ * stored before processing.
+ *
+ * The supplied MPI-1 code generates the input stream {A} on all processors
+ * and the global table has been distributed as uniformly as possible to
+ * balance the workload and minimize any Amdahl fraction.  This code does not
+ * exploit "look-ahead".  Addresses are sent to the appropriate processor
+ * where the table entry resides as soon as each address is calculated.
+ * Updates are performed as addresses are received.  Each message is limited
+ * to a single 64 bit long integer containing element ai from {A}.
+ * Local offsets for T[ ] are extracted by the destination processor.
+ *
+ * If the number of processors is equal to a power of two, then the global
+ * table can be distributed equally over the processors.  In addition, the
+ * processor number can be determined from that portion of the input stream
+ * that identifies the address into the global table by masking off log2(p)
+ * bits in the address.
+ *
+ * If the number of processors is not equal to a power of two, then the global
+ * table cannot be equally distributed between processors.  In the MPI-1
+ * implementation provided, there has been an attempt to minimize the differences
+ * in workloads and the largest difference in elements of T[ ] is one.  The
+ * number of values in the input stream generated by each processor will be
+ * related to the number of global table entries on each processor.
+ *
+ * The MPI-1 version of RandomAccess treats the potential instance where the
+ * number of processors is a power of two as a special case, because of the
+ * significant simplifications possible because processor location and local
+ * offset can be determined by applying masks to the input stream values.
+ * The non power of two case uses an integer division to determine the processor
+ * location.  The integer division will be more costly in terms of machine
+ * cycles to perform than the bit masking operations
+ *
+ * For additional information on the GUPS metric, the HPCchallenge RandomAccess
+ * Benchmark,and the rules to run RandomAccess or modify it to optimize
+ * performance -- see http://icl.cs.utk.edu/hpcc/
+ */
+
+
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <assert.h>
+#include <time.h>
+#include <sys/time.h>
+#include <string.h>
+#include <pthread.h>
```

Comment:
Copy paste from rdm_one_sided (which I used copy the gnix provider setup). I'll remove it.